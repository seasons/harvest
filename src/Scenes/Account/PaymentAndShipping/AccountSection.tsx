import { Box, Sans, Spacer, Separator, Flex } from "App/Components"
import React from "react"
import { color } from "App/utils"
import { TouchableOpacity } from "react-native"

export const AccountSection: React.FC<{ title: string; value: string | [string]; onEdit: () => void }> = ({
  title,
  value,
  onEdit,
}) => {
  return (
    <Box key={title} px={2}>
      <Flex flexDirection="row" justifyContent={!!onEdit ? "space-between" : "flex-start"} alignItems="center">
        <Sans size="4">{title}</Sans>
        {!!onEdit && (
          <TouchableOpacity onPress={onEdit}>
            <Sans size="4" style={{ textDecorationLine: "underline" }}>
              Edit
            </Sans>
          </TouchableOpacity>
        )}
      </Flex>
      <Box mb={1} />
      <Separator color={color("black10")} />
      <Box mb={1} />
      {Array.isArray(value) ? (
        value.map((text) => (
          <Sans key={text} size="4" color="black50">
            {text}
          </Sans>
        ))
      ) : (
        <Sans size="4" color="black50">
          {value}
        </Sans>
      )}
      <Spacer mb={4} />
    </Box>
  )
}
