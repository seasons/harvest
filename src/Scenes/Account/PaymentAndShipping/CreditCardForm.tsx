import React, { Component } from "react"
import { ScrollView, StyleSheet, View } from "react-native"
import stripe, { PaymentCardTextField, StripeToken } from "tipsi-stripe"
import { Box, Button, Sans } from "App/Components"

export interface PaymentCardTextFieldParams {
  number: string
  expMonth: string
  expYear: string
  cvc: string
  name?: string
  addressLine1?: string
  addressLine2?: string
  addressCity?: string
  addressState?: string
  addressZip?: string
}

export interface StripeToken {
  tokenId: string
  created: number
  livemode: 1 | 0
  card: any
  bankAccount: any
  extra: any
}

interface CreditCardFormProps {
  navigation?: NavigationType
  params?: PaymentCardTextFieldParams
  onSubmit: (t: StripeToken, p: PaymentCardTextFieldParams) => void
}

interface CreditCardFormState {
  valid: boolean
  params: PaymentCardTextFieldParams
  isLoading: boolean
  isError: boolean
}

export class CreditCardForm extends Component<CreditCardFormProps, CreditCardFormState> {
  private paymentInfo: PaymentCardTextField

  constructor(props) {
    super(props)

    this.paymentInfo = (React as any).createRef()
    this.state = { valid: null, params: { ...this.props.params }, isLoading: false, isError: false }
  }

  componentDidMount() {
    if (this.paymentInfo.value) {
      this.paymentInfo.value.setParams(this.state.params)
      this.paymentInfo.value.focus()
    }
  }

  handleFieldParamsChange = (valid, params: PaymentCardTextFieldParams) => {
    this.setState({ valid, params })
  }

  tokenizeCardAndSubmit = async () => {
    this.setState({ isLoading: true, isError: false })

    const { params } = this.state

    try {
      const token = await stripe.createTokenWithCard({ ...params })
      this.props.onSubmit(token, this.state.params)
      this.setState({ isLoading: false })
      this.props.navigation.pop()
    } catch (error) {
      console.warn("CreditCardForm.tsx", error)
      this.setState({ isError: true, isLoading: false })
    }
  }

  render() {
    const buttonComponent = (
      <Box m={4}>
        <Button
          disabled={!this.state.valid}
          width={100}
          onPress={this.state.valid ? () => this.tokenizeCardAndSubmit() : null}
        >
          Add credit card
        </Button>
      </Box>
    )

    // const styles = StyleSheet.create({
    //   field: {
    //     fontFamily: Fonts.GaramondRegular,
    //     height: 40,
    //     fontSize: theme.fontSizes[3],
    //     width: "100%",
    //     borderColor: this.state.isError ? theme.colors.red100 : theme.colors.purple100,
    //     borderWidth: 1,
    //     borderRadius: 0,
    //   },
    // })

    const errorText = "There was an error. Please try again."

    return (
      <>
        <ScrollView scrollEnabled={false}>
          <Box m={4}>
            <Sans size="3" mb={2}>
              Card Information
            </Sans>
            <PaymentCardTextField
              ref={this.paymentInfo}
              onParamsChange={this.handleFieldParamsChange}
              numberPlaceholder="Card number"
              expirationPlaceholder="MM/YY"
              cvcPlaceholde="CVC"
            />
            {this.state.isError && (
              <Sans size="2" mt={3} color="red100">
                {errorText}
              </Sans>
            )}
            <Sans mt="6" size="3" color="black60" textAlign="center">
              Registration is free.
              {"\n"}A credit card is required to bid. Artsy will never charge this card without your permission, and you
              are not required to use this card to pay if you win.
            </Sans>
          </Box>
        </ScrollView>
      </>
    )
  }
}
