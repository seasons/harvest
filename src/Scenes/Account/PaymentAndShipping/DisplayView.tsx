import { Box, Sans, Separator, Spacer } from "App/Components"

import React from "react"

const sections = [
  { title: "Delivery address", textLines: ["138 Mulberry St", "New York, New York 1234"] },
  { title: "Billing address", textLines: ["138 Mulberry St", "New York, New York 1234"] },
  { title: "Payment info", textLines: ["Visa 1234"] },
  { title: "Phone number", textLines: ["555 555 5555"] },
]

export class DisplayView extends React.Component {
  render() {
    return (
      <>
        {sections.map(section => {
          return (
            <Box mb={4} key={section.title}>
              <Sans size="2">{section.title}</Sans>
              <Spacer mb={2} />
              <Separator />
              <Spacer mb={2} />
              {section.textLines.map((string, index) => (
                <Sans color="gray" size="2" key={section.title + index}>
                  {string}
                </Sans>
              ))}
            </Box>
          )
        })}
      </>
    )
  }
}
