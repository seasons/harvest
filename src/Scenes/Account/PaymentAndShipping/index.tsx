export { EditPaymentAndShipping } from "./EditPaymentAndShipping"
export { PaymentAndShipping } from "./PaymentAndShipping"
