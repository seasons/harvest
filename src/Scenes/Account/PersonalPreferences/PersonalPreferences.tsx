import { Box, Container, FixedBackArrow, Sans, Separator, Spacer } from "App/Components"
import { Loader } from "App/Components/Loader"
import { color } from "App/Utils"
import gql from "graphql-tag"
import { DateTime } from "luxon"
import React, { useEffect, useState } from "react"
import { useQuery } from "react-apollo"
import { FlatList } from "react-native"
import { NavigationParams, NavigationScreenProp, NavigationState } from "react-navigation"

const GET_PREFERENCES = gql`
  query getUser {
    me {
      customer {
        detail {
          phoneNumber
          birthday
          height
          weight
          bodyType
          averageSpend
          averageTopSize
          averageWaistSize
          profession
          partyFrequency
          travelFrequency
          shoppingFrequency
          style
          phoneOS
          commuteStyle
          preferredPronouns
          averagePantLength
        }
      }
    }
  }
`

export const AccountSection: React.FC<{ title: string; value: string | [string] }> = ({ title, value }) => {
  return (
    <Box key={title} px={2}>
      <Sans size="2">{title}</Sans>
      <Box mb={1} />
      <Separator color={color("gray")} />
      <Box mb={1} />
      {Array.isArray(value) ? (
        value.map(text => (
          <Sans key={text} size="2" color="gray">
            {text}
          </Sans>
        ))
      ) : (
        <Sans size="2" color="gray">
          {value}
        </Sans>
      )}
      <Spacer mb={2} />
    </Box>
  )
}

export const PersonalPreferences: React.FC<{ navigation: NavigationScreenProp<NavigationState, NavigationParams> }> = ({
  navigation,
}) => {
  const insets = useSafeArea()
  const [sections, setSections] = useState([])
  const { loading, error, data } = useQuery(GET_PREFERENCES)

  useEffect(() => {
    if (data && data.me && data.me.customer && data.me.customer.detail) {
      const sectionsArray = []
      const details = data.me.customer.detail

      if (details.birthday) {
        const birthdayAsDate = DateTime.fromISO(details.birthday)
          .toUTC()
          .toLocaleString(DateTime.DATE_FULL)
        sectionsArray.push({ title: "Birthday", value: birthdayAsDate })
      }

      if (details.preferredPronouns) {
        sectionsArray.push({ title: "Preferred pronouns", value: details.preferredPronouns })
      }

      if (details.height) {
        const feet = Math.floor(details.height / 12)
        const inches = details.height % 12
        sectionsArray.push({ title: "Height", value: `${feet} ft ${inches} in` })
      }

      if (details.weight) {
        sectionsArray.push({ title: "Weight", value: details.weight })
      }

      if (details.bodyType) {
        sectionsArray.push({ title: "Body type", value: details.bodyType })
      }

      if (details.averageTopSize) {
        sectionsArray.push({ title: "Average top size", value: details.averageTopSize })
      }

      if (details.averageWaistSize) {
        sectionsArray.push({ title: "Average waist size", value: `${details.averageWaistSize} in` })
      }

      if (details.averagePantLength) {
        sectionsArray.push({ title: "Average pant length", value: `${details.averagePantLength} in` })
      }

      // page 2

      if (details.profession) {
        sectionsArray.push({ title: "Profession", value: details.profession })
      }

      if (details.commuteStyle) {
        sectionsArray.push({ title: "Commute style", value: details.commuteStyle })
      }

      if (details.partyFrequency) {
        sectionsArray.push({ title: "Party frequency", value: details.partyFrequency })
      }

      if (details.travelFrequency) {
        sectionsArray.push({ title: "Travel frequency", value: details.travelFrequency })
      }

      if (details.shoppingFrequency) {
        sectionsArray.push({ title: "Shopping frequency", value: details.shoppingFrequency })
      }

      if (details.averageSpend) {
        sectionsArray.push({ title: "Spending average", value: details.averageSpend })
      }

      if (details.style) {
        sectionsArray.push({ title: "Style", value: details.style })
      }

      if (details.phoneOS) {
        sectionsArray.push({ title: "Phone type", value: details.phoneOS })
      }

      setSections(sectionsArray)
    }
  }, [data])

  if (loading) {
    return <Loader />
  }

  if (error) {
    console.log("error PersonalPreferences.tsx: ", error)
    return null
  }

  const renderItem = item => {
    return <AccountSection title={item.title} value={item.value} />
  }

  return (
    <Container>
      <Box mt={insets.top}>
        <FixedBackArrow navigation={navigation} />
        <FlatList
          data={sections}
          ListHeaderComponent={() => (
            <Box px={2} mt={4}>
              <Spacer mb={80} />
              <Sans size="3">Personal preferences</Sans>
              <Spacer mb={3} />
            </Box>
          )}
          keyExtractor={item => item.title}
          renderItem={({ item }) => renderItem(item)}
        />
      </Box>
    </Container>
  )
}
