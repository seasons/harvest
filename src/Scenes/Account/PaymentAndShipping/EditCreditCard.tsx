import { Box, Button, Container, FixedBackArrow, Flex, Sans, Separator, Spacer, TextInput } from "App/Components"
import { Schema as NavigationSchema } from "App/Navigation"
import { usePopUpContext } from "App/Navigation/ErrorPopUp/PopUpContext"
import { color } from "App/utils/color"
import { space } from "App/utils/space"
import { Schema as TrackSchema, screenTrack, useTracking } from "App/utils/track"
import { CheckCircled } from "Assets/svgs/CheckCircled"
import { String } from "aws-sdk/clients/augmentedairuntime"
import React, { useEffect, useState } from "react"
import { Dimensions, KeyboardAvoidingView, ScrollView } from "react-native"
import styled from "styled-components"
import stripe, { PaymentCardTextField } from "tipsi-stripe"
import { useMutation, useQuery } from "@apollo/client"
import { useNotificationBarContext } from "@seasons/eclipse"
import * as Sentry from "@sentry/react-native"

import { PAYMENT_UPDATE } from "./PaymentAndShipping"
import { GET_PAYMENT_DATA } from "./queries"

interface BillingAddress {
  street1: string
  street2: string
  city: string
  state: string
  postal_code: String
}

const windowDimensions = Dimensions.get("window")
const windowWidth = windowDimensions.width
const singleButtonWidth = windowWidth - 42
const buttonWidth = singleButtonWidth / 2

const PaymentForm = ({ setBillingAddress, isMutating, handleFieldParamsChange, billingAddress }) => {
  const address1 = billingAddress?.address1
  const address2 = billingAddress?.address2
  const city = billingAddress?.city
  const state = billingAddress?.state
  const zipCode = billingAddress?.zipCode

  return (
    <Box px={2} pb={150}>
      <Spacer mb={4} />
      <Sans size="4">Card information</Sans>
      <Spacer mb={3} />
      <Flex flexDirection="row" flexWrap="nowrap" justifyContent="flex-start">
        <Sans color="black50" size="3">
          Card number, Expiration date & CIV
        </Sans>
      </Flex>
      <PaymentField
        cursorColor={color("black50")}
        textErrorColor={color("black50")}
        placeholderColor={color("black50")}
        numberPlaceholder="16 digits"
        expirationPlaceholder="MM / YY"
        cvcPlaceholder="3-digits"
        disabled={isMutating}
        onParamsChange={handleFieldParamsChange}
      />
      <Separator />
      <Spacer mb={4} />
      <Sans size="4">Billing address</Sans>
      <Spacer mb={2} />
      <TextInput
        currentValue={address1}
        placeholder="Address 1"
        onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, address1: text })}
      />
      <Spacer mb={2} />
      <Flex flexDirection="row" flexWrap="nowrap" justifyContent="space-between">
        <TextInput
          currentValue={address2}
          placeholder="Address 2"
          style={{ flex: 1 }}
          onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, address2: text })}
        />
        <Spacer ml={1} />
        <TextInput
          currentValue={zipCode}
          placeholder="Zip code"
          style={{ flex: 1 }}
          onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, zipCode: text })}
        />
      </Flex>
      <Spacer mb={2} />
      <Flex flexDirection="row" flexWrap="nowrap" justifyContent="space-between">
        <TextInput
          currentValue={city}
          placeholder="City"
          style={{ flex: 1 }}
          onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, city: text })}
        />
        <Spacer ml={1} />
        <TextInput
          currentValue={state}
          placeholder="State"
          style={{ flex: 1 }}
          onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, state: text })}
        />
      </Flex>
    </Box>
  )
}

export const EditCreditCard: React.FC<{
  navigation: any
  route: any
}> = screenTrack()(({ route, navigation }) => {
  let paymentPlan = route?.params?.paymentPlan
  let routeBillingAddress: BillingAddress = route?.params?.billingAddress
  const { data } = useQuery(GET_PAYMENT_DATA, { skip: !!paymentPlan })
  if (!paymentPlan) {
    paymentPlan = data?.me?.customer?.paymentPlan
    routeBillingAddress = data?.me?.customer?.billingInfo
  }

  const tracking = useTracking()

  const { hideNotificationBar } = useNotificationBarContext()
  const { showPopUp, hidePopUp } = usePopUpContext()
  const [isMutating, setIsMutating] = useState(false)
  const [onPaymentUpdate, setOnPaymentUpdate] = useState(false)
  const [cardNumber, setCardNumber] = useState("")
  const [expMonth, setExpMonth] = useState("")
  const [expYear, setExpYear] = useState("")
  const [cvc, setCvc] = useState("")
  const [billingAddress, setBillingAddress] = useState({
    address1: "",
    address2: "",
    city: "",
    state: "",
    zipCode: "",
  })

  const address1 = billingAddress?.address1
  const address2 = billingAddress?.address2
  const city = billingAddress?.city
  const state = billingAddress?.state
  const zipCode = billingAddress?.zipCode

  useEffect(() => {
    if (routeBillingAddress) {
      const address1 = routeBillingAddress?.street1 || ""
      const address2 = routeBillingAddress?.street2 || ""
      const city = routeBillingAddress?.city || ""
      const state = routeBillingAddress?.state || ""
      const zipCode = routeBillingAddress?.postal_code || ""

      setBillingAddress({
        address1,
        address2,
        city,
        state,
        zipCode,
      })
    }
  }, [routeBillingAddress])

  const [updateCreditCard] = useMutation(PAYMENT_UPDATE, {
    refetchQueries: [
      {
        query: GET_PAYMENT_DATA,
      },
    ],
    onCompleted: () => {
      setIsMutating(false)
      setOnPaymentUpdate(true)
    },
    onError: (err) => {
      console.log("Error ChoosePlanPane.tsx", err)
      Sentry.captureException(JSON.stringify(err))
      const popUpData = {
        title: "Oops! Try again!",
        note: "There was an issue updating your payment method. Please retry or contact us.",
        buttonText: "Close",
        onClose: hidePopUp,
      }
      showPopUp(popUpData)
      setIsMutating(false)
    },
  })

  const handleFieldParamsChange = (valid, values) => {
    const { number, expMonth, expYear, cvc } = values

    setCardNumber(number)
    setExpMonth(expMonth)
    setExpYear(expYear)
    setCvc(cvc)
  }

  const handleOnSubmit = async () => {
    if (isMutating) {
      return
    }
    setIsMutating(true)

    tracking.trackEvent({
      actionName: TrackSchema.ActionNames.AddCreditCardTapped,
      actionType: TrackSchema.ActionTypes.Tap,
    })

    if (!address1 || !city || !state || !zipCode) {
      showPopUp({
        title: "Your billing address is incomplete",
        note: "Please complete your billing address before updating your card number.",
        buttonText: "Close",
        onClose: hidePopUp,
      })
      setIsMutating(false)
    } else {
      try {
        const params = {
          // mandatory
          number: cardNumber,
          expMonth,
          expYear,
          cvc,
          // optional
          addressLine1: address1,
          addressLine2: address2,
          addressCity: city,
          addressState: state,
          // addressCountry: "Test Country",
          addressZip: zipCode,
        }
        const token = await stripe.createTokenWithCard(params)
        updateCreditCard({
          variables: {
            planID: paymentPlan.planID,
            token,
            tokenType: "card",
          },
          awaitRefetchQueries: true,
        })
        // You should complete the operation by calling
        stripe.completeApplePayRequest()
        hideNotificationBar()
      } catch (error) {
        const popUpData = {
          title: "Oops! Try again!",
          note: "Your card is invalid.",
          buttonText: "Close",
          onClose: hidePopUp,
        }
        showPopUp(popUpData)
        console.log("error", error)
        setIsMutating(false)
      }
    }
  }

  return (
    <Container insetsBottom={false}>
      {!onPaymentUpdate && (
        <FixedBackArrow
          navigation={navigation}
          onPress={() => {
            navigation.navigate(NavigationSchema.StackNames.AccountStack, {
              screen: NavigationSchema.PageNames.PaymentAndShipping,
            })
          }}
          variant="whiteBackground"
        />
      )}
      <ScrollView style={{ flex: 1 }}>
        <Spacer mt={100} />
        <Box px={2}>
          {onPaymentUpdate && (
            <>
              <CheckCircled backgroundColor={color("green100")} />
              <Spacer mb={2} />
            </>
          )}
          <Sans size="7">{onPaymentUpdate ? "Payment method updated" : "Update your payment method"}</Sans>
        </Box>
        {!onPaymentUpdate && (
          <PaymentForm
            setBillingAddress={setBillingAddress}
            isMutating={isMutating}
            handleFieldParamsChange={handleFieldParamsChange}
            billingAddress={billingAddress}
          />
        )}
      </ScrollView>
      <KeyboardAvoidingView behavior="padding" keyboardVerticalOffset={space(2)} style={{ bottom: space(2) }}>
        {onPaymentUpdate ? (
          <Flex flexDirection="row" flexWrap="nowrap" justifyContent="center">
            <Button
              disabled={isMutating}
              variant="primaryBlack"
              size="large"
              width={singleButtonWidth}
              onPress={() => {
                navigation.navigate(NavigationSchema.StackNames.AccountStack, {
                  screen: NavigationSchema.PageNames.PaymentAndShipping,
                })
              }}
            >
              Close
            </Button>
          </Flex>
        ) : (
          <Flex flexDirection="row" flexWrap="nowrap" justifyContent="center">
            <Button
              disabled={isMutating}
              variant="primaryWhite"
              width={buttonWidth}
              onPress={() => {
                navigation.goBack()
              }}
            >
              Go back
            </Button>
            <Spacer ml={1} />
            <Button width={buttonWidth} block onPress={handleOnSubmit} loading={isMutating} disabled={isMutating}>
              Submit
            </Button>
          </Flex>
        )}
      </KeyboardAvoidingView>
    </Container>
  )
})

const PaymentField = styled(PaymentCardTextField)`
  color: ${color("black100")};
  font-size: 16px;
`
