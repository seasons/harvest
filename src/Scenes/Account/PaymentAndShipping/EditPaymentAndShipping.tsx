import gql from "graphql-tag"
import React, { useEffect, useState } from "react"
import { Dimensions, FlatList } from "react-native"
import { useSafeArea } from "react-native-safe-area-context"
import { Box, Button, Container, Flex, FixedBackArrow, PopUp, Radio, Sans, Spacer, TextInput } from "App/Components"
import { get } from "lodash"
import { chargebeeUpdatePaymentPage_chargebeeUpdatePaymentPage } from "src/generated/chargebeeUpdatePaymentPage"
import { GET_PAYMENT_DATA } from "./PaymentAndShipping"
import {
  GetUserPaymentData_me_customer_billingInfo,
  GetUserPaymentData_me_customer_detail_shippingAddress
} from "src/generated/getUserPaymentData"
import { useMutation, useQuery } from "react-apollo"

const UPDATE_PAYMENT_AND_SHIPPING = gql`
  mutation updatePaymentAndShipping($billingAddress: AddressInput!, $shippingAddress: AddressInput!) {
    updatePaymentAndShipping(billingAddress: $billingAddress, shippingAddress: $shippingAddress) 
  }
`

const BILLING_ADDRESS = "Billing address"
const DELIVERY_ADDRESS = "Delivery address"
const EDIT_BILLING_INFO = "Edit billing info"
const FINISH_BUTTONS = "Finish buttons"

export const EditPaymentAndShipping: React.FC<{
  navigation: any
  route: any
}> = ({ navigation, route }) => {
  const billingInfo: GetUserPaymentData_me_customer_billingInfo = get(route, "params.billingInfo")
  const chargebeeUpdatePaymentHostedPage: chargebeeUpdatePaymentPage_chargebeeUpdatePaymentPage = get(route, "params.chargebeeUpdatePaymentHostedPage")
  const currentShippingAddress: GetUserPaymentData_me_customer_detail_shippingAddress = get(route, "params.shippingAddress")

  const [showError, setShowError] = useState(false)
  const [isMutating, setIsMutating] = useState(false)
  const [shippingAddress, setShippingAddress] = useState({
    address1: currentShippingAddress?.address1 || "",
    address2: currentShippingAddress?.address2 || "",
    city: currentShippingAddress?.city || "",
    state: currentShippingAddress?.state || "",
    zipCode: currentShippingAddress?.zipCode || ""
  })

  const [billingAddress, setBillingAddress] = useState({
    address1: billingInfo?.street1 || "",
    address2: billingInfo?.street2 || "",
    city: billingInfo?.city || "",
    state: billingInfo?.state || "",
    zipCode: billingInfo?.postal_code || ""
  })
  const [sameAsDeliveryRadioSelected, setSameAsDeliveryRadioSelected] = useState(false)

  const {
    address1: shippingAddress1,
    address2: shippingAddress2,
    city: shippingCity,
    state: shippingState,
    zipCode: shippingZipCode,
  } = shippingAddress

  const {
    address1: billingAddress1,
    address2: billingAddress2,
    city: billingCity,
    state: billingState,
    zipCode: billingZipCode,
  } = billingAddress

  const [updatePaymentAndShipping] = useMutation(UPDATE_PAYMENT_AND_SHIPPING, {
    onError: error => {
      setShowError(true)
      console.log("error EditView.tsx: ", error)
    },
  })

  const handleSameAsDeliveryAddress = () => {
    if (sameAsDeliveryRadioSelected) {
      setBillingAddress({
        address1: "",
        address2: "",
        city: "",
        state: "",
        zipCode: "",
      })
    } else {
      setBillingAddress({
        address1: shippingAddress1,
        address2: shippingAddress2,
        city: shippingCity,
        state: shippingState,
        zipCode: shippingZipCode,
      })
      console.log("SET BILLING CITY:", shippingCity)
    }
    setSameAsDeliveryRadioSelected(!sameAsDeliveryRadioSelected)
  }

  const handleSaveBtnPressed = async () => {
    setIsMutating(true)
    const result = await updatePaymentAndShipping({
      variables: {
        billingAddress: {
          city: billingCity,
          postalCode: billingZipCode,
          state: billingState,
          street1: billingAddress1,
          street2: billingAddress2
        },
        shippingAddress: {
          city: shippingCity,
          postalCode: shippingZipCode,
          state: shippingState,
          street1: shippingAddress1,
          street2: shippingAddress2
        }
      },
      refetchQueries: [
        {
          query: GET_PAYMENT_DATA
        }
      ]
    })
    console.log("UPDATED INFO:", result)
    setIsMutating(false)
    if (result) {
      navigation.pop()
    }
  }

  const handleEditBillingInfoBtnPressed = () => {
    if (chargebeeUpdatePaymentHostedPage.url) {
      navigation.navigate("Webview", {
        uri: chargebeeUpdatePaymentHostedPage.url,
        variant: "whiteBackground"
      })
    }
  }

  const handleCancelBtnPressed = () => {
    navigation.pop()
  }

  const sections = [DELIVERY_ADDRESS, BILLING_ADDRESS, EDIT_BILLING_INFO, FINISH_BUTTONS]

  const screenWidth = Dimensions.get("window").width
  const buttonWidth = (screenWidth - 40) / 2
  const renderItem = ({ item: section }) => {
    switch (section) {
      case DELIVERY_ADDRESS:
        return (
          <>
            <Sans size="1">{DELIVERY_ADDRESS}</Sans>
            <Spacer mb={2} />
            <TextInput currentValue={shippingAddress1} placeholder="Address 1" onChangeText={(inputKey, text) => setShippingAddress({ ...shippingAddress, address1: text })} />
            <Spacer mb={2} />
            <Flex flexDirection="row" flexWrap="nowrap" justifyContent="center">
              <TextInput currentValue={shippingAddress2} placeholder="Address 2" style={{ flex: 1 }} onChangeText={(inputKey, text) => setShippingAddress({ ...shippingAddress, address2: text })} />
              <Spacer ml={1} />
              <TextInput currentValue={shippingZipCode} placeholder="Zipcode" style={{ flex: 1 }} onChangeText={(inputKey, text) => setShippingAddress({ ...shippingAddress, zipCode: text })} />
            </Flex>
            <Spacer mb={2} />
            <Flex flexDirection="row" flexWrap="nowrap" justifyContent="space-between">
              <TextInput currentValue={shippingCity} placeholder="City" style={{ flex: 1 }} onChangeText={(inputKey, text) => { setShippingAddress({ ...shippingAddress, city: text }) }} />
              <Spacer ml={1} />
              <TextInput currentValue={shippingState} placeholder="State" style={{ flex: 1 }} onChangeText={(inputKey, text) => { setShippingAddress({ ...shippingAddress, state: text }) }} />
            </Flex>
          </>
        )
      case BILLING_ADDRESS:
        return (
          <>
            <Sans size="1">{BILLING_ADDRESS}</Sans>
            <Spacer mb={2} />
            <Radio
              borderRadius={4}
              selected={sameAsDeliveryRadioSelected}
              onSelect={handleSameAsDeliveryAddress}
              label="Same as Delivery Address"
              labelSize="1"
            />
            <Spacer mb={2} />
            <TextInput currentValue={billingAddress1} placeholder="Address 1" onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, address1: text })} />
            <Spacer mb={2} />
            <Flex flexDirection="row" flexWrap="nowrap" justifyContent="space-between">
              <TextInput currentValue={billingAddress2} placeholder="Address 2" style={{ flex: 1 }} onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, address2: text })} />
              <Spacer ml={1} />
              <TextInput currentValue={billingZipCode} placeholder="Zipcode" style={{ flex: 1 }} onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, zipCode: text })} />
            </Flex>
            <Spacer mb={2} />
            <Flex flexDirection="row" flexWrap="nowrap" justifyContent="space-between">
              <TextInput currentValue={billingCity} placeholder="City" style={{ flex: 1 }} onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, city: text })} />
              <Spacer ml={1} />
              <TextInput currentValue={billingState} placeholder="State" style={{ flex: 1 }} onChangeText={(inputKey, text) => setBillingAddress({ ...billingAddress, state: text })} />
            </Flex>
          </>
        )
      case EDIT_BILLING_INFO:
        return (
          <Flex flexDirection="row" justifyContent="space-between">
            <Button variant="secondaryBlack" size="large" width={buttonWidth} onPress={handleEditBillingInfoBtnPressed}>
              Edit Billing Info
            </Button>
          </Flex>
        )
      case FINISH_BUTTONS:
        return (
          <Flex flexDirection="row" justifyContent="space-between">
            <Button variant="primaryWhite" size="large" width={buttonWidth} onPress={handleCancelBtnPressed}>
              Cancel
            </Button>
            <Button loading={isMutating} variant="secondaryBlack" size="large" width={buttonWidth} onPress={handleSaveBtnPressed}>
              Save
            </Button>
          </Flex>
        )
      default:
        return null
    }
  }

  const popUpData = {
    buttonText: "Got it",
    note: "Make sure your shipping and billing address are valid.",
    title: "Something went wrong!",
    onClose: () => setShowError(false)
  }

  const insets = useSafeArea()
  return (
    <Container insetsBottom={0}>
      <FixedBackArrow navigation={navigation} variant="whiteBackground" />
      <Box px={2}>
        <FlatList
          data={sections}
          ListHeaderComponent={() => (
            <Box mt={insets.top}>
              <Spacer mb={2} />
              <Sans size="3">Payment & Shipping</Sans>
              <Spacer mb={6} />
            </Box>
          )}
          ItemSeparatorComponent={() => <Spacer mb={6} />}
          keyExtractor={(item, index) => item + String(index)}
          renderItem={renderItem}
          ListFooterComponent={() => <Spacer mb={2} />}
          showsVerticalScrollIndicator={false}
        />
      </Box>
      <PopUp data={popUpData} show={showError} insetsBottom={true} />
    </Container>
  )
}
