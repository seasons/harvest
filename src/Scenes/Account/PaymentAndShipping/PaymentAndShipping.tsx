import { Box, Button, Flex, Theme } from "App/Components"
import { DisplayView } from "./DisplayView"
import { EditView } from "./EditView"
import React from "react"
import { SafeAreaView } from "react-native"
import { space } from "App/Utils"
import styled from "styled-components/native"

interface Props {}

interface State {
  renderEditView: boolean
}

export class PaymentAndShipping extends React.Component<Props, State> {
  state = {
    renderEditView: false,
  }

  toggleView = () => {
    this.setState({
      renderEditView: !this.state.renderEditView,
    })
  }

  render() {
    const { renderEditView } = this.state

    return (
      <SafeAreaView style={{ flex: 1 }}>
        <Theme>
          <Flex style={{ flex: 1 }}>
            {renderEditView ? <EditView {...this.props} /> : <DisplayView {...this.props} />}
            <FixedButtonWrapper px={2}>
              <Button variant={renderEditView ? "primaryDark" : "secondaryDark"} onPress={() => this.toggleView()}>
                {renderEditView ? "Done" : "Edit"}
              </Button>
            </FixedButtonWrapper>
          </Flex>
        </Theme>
      </SafeAreaView>
    )
  }
}

const FixedButtonWrapper = styled(Box)`
  position: absolute;
  bottom: ${space(1)};
`
