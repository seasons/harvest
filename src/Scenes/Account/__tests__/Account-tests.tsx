import React from "react"
import { PaymentAndShipping } from ".."
import { mount } from "enzyme"
import { MockedProvider } from "@apollo/react-testing"
import { GetPaymentDataFixture } from "./__fixtures__/AccountFixture"
import { waitForLoad } from "App/testUtils/waitForLoad"
import { AccountSection } from "../PersonalPreferences/PersonalPreferences"
import { GET_PAYMENT_DATA } from "../PaymentAndShipping/PaymentAndShipping"

describe("PaymentAndShipping", () => {
  const paymentAndShippingMocks = [
    {
      request: {
        query: GET_PAYMENT_DATA,
      },
      result: {
        data: {
          ...GetPaymentDataFixture,
        },
      },
    },
  ]

  it("renders properly", async () => {
    const component = mount(
      <MockedProvider mocks={paymentAndShippingMocks} addTypename={false}>
        <PaymentAndShipping navigation={null} />
      </MockedProvider>
    )

    await waitForLoad(component)

    expect(component.find(AccountSection).length).toEqual(4)
    expect(component.text()).toContain(
      "Payment & ShippingShipping address1 Bedford AvenueApt 5ABROOKLYN, New York, 11211Billing address1 Bedford AvenueBrooklynBrooklyn, New York, 11211Payment info4242Phone number123-456-1234"
    )
    expect(component).toMatchSnapshot()
  })
})
