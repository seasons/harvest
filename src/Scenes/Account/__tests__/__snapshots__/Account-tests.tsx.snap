// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentAndShipping renders properly 1`] = `ReactWrapper {}`;
